### PR TITLE
Enums/Measures

### PR DESCRIPTION
**Pull request by ranjitjhala**
Support for enums-with-measures; e.g. see 

- /tests/pos/enums/opt00.rs
- /tests/pos/enums/list00.rs

I still want to add some neg-tests but want to make this PR to get feedback early.

### CODE REVIEW

File: flux-middle/src/ty/subst.rs [hunk 1]
```diff
@@ -80,13 +80,13 @@ impl TypeFolder for FVarFolder<'_> {
     }
 }
 
-pub(crate) struct BVarFolder<'a> {
+pub struct BVarFolder<'a> {
```
Comment (nilehmann) [inline flux-middle/src/ty/subst.rs:83]: This `pub(crate)` is on purpose to try enforcing the `Binders` abstraction.

```suggestion
pub(crate) struct BVarFolder<'a> {
```

File: flux-middle/src/ty/subst.rs [hunk 2]
```diff
@@ -80,13 +80,13 @@ impl TypeFolder for FVarFolder<'_> {
...
     outer_binder: DebruijnIndex,
     exprs: &'a [Expr],
 }
 
 impl<'a> BVarFolder<'a> {
-    pub(crate) fn new(exprs: &'a [Expr]) -> BVarFolder<'a> {
+    pub fn new(exprs: &'a [Expr]) -> BVarFolder<'a> {
```
Comment (nilehmann) [inline flux-middle/src/ty/subst.rs:89]:
```suggestion
    pub(crate) fn new(exprs: &'a [Expr]) -> BVarFolder<'a> {
```

**Review by nilehmann — CHANGES_REQUESTED** [review]
This is very exciting!

File: flux-typeck/src/type_env/paths_tree.rs [hunk 3]
```diff
@@ -46,6 +48,89 @@ impl LookupResult {
...
+///     3. *Assert* the constraint `i == j'...`
+
+fn enum_constraint(scrutinee: &flux_middle::ty::Ty, exprs: &[flux_middle::ty::Expr]) -> Expr {
+    match scrutinee.kind() {
+        TyKind::Indexed(_, ixs) => {
+            assert_eq!(ixs.len(), exprs.len());
+            Expr::and(iter::zip(ixs, exprs).map(|(i, e)| Expr::eq(i.expr.clone(), e.clone())))
+        }
+        TyKind::Constr(e, ty) => Expr::and2(e.clone(), enum_constraint(ty, exprs)),
+        _ => panic!("unexpected: enum_constraint {scrutinee:?}"),
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-typeck/src/type_env/paths_tree.rs:94]: Yes, may be better to do this as the use cases arise.

File: flux-middle/src/ty/mod.rs [hunk 4]
```diff
@@ -527,10 +529,14 @@ impl Expr {
             .clone()
     }
 
+    pub fn and2(e1: Expr, e2: Expr) -> Expr {
+        Expr::binary_op(BinOp::And, e1, e2)
+    }
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-middle/src/ty/mod.rs:534]: nice, let me use the array instead!

File: flux-typeck/src/type_env/paths_tree.rs [hunk 5]
```diff
@@ -46,6 +48,89 @@ impl LookupResult {
...
+///     * `x.fld : T[A := t ..]<i := e...>`
+/// i.e. by substituting the type and value indices using the types and values from `x`.
+
+fn downcast_struct(
+    genv: &GlobalEnv,
+    def_id: DefId,
+    variant_idx: VariantIdx,
+    substs: &[flux_middle::ty::Ty],
+    exprs: &[flux_middle::ty::Expr],
+) -> Vec<flux_middle::ty::Ty> {
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-typeck/src/type_env/paths_tree.rs:66]: done!

File: flux-typeck/src/type_env/paths_tree.rs [hunk 6]
```diff
@@ -46,6 +48,89 @@ impl LookupResult {
...
+/// In contrast (w.r.t. `struct`) downcast on `enum` works as follows.
+/// Given
+///     * a "place" `x : T[i..]`
+///     * a "variant" of type `forall z..,(y:t...) => T[j...]`
+/// We want `downcast` to return a vector of types _and an assertion_ by
+///     1. *Instantiate* the type to fresh names `z'...` to get `(y:t'...) => T[j'...]`
+///     2. *Unpack* the fields using `y:t'...`
+///     3. *Assert* the constraint `i == j'...`
+
+fn enum_constraint(scrutinee: &flux_middle::ty::Ty, exprs: &[flux_middle::ty::Expr]) -> Expr {
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-typeck/src/type_env/paths_tree.rs:87]: done!

File: flux-typeck/src/type_env/paths_tree.rs [hunk 7]
```diff
@@ -331,6 +424,13 @@ enum NodeKind {
     Uninit,
 }
 
+fn strip_constr(ty: &Ty) -> &Ty {
+    match ty.kind() {
+        TyKind::Indexed(_, _) => ty,
+        TyKind::Constr(_, ty1) => strip_constr(ty1),
+        _ => todo!(),
+    }
+}
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-typeck/src/type_env/paths_tree.rs:429]: good catch!

File: flux-typeck/src/type_env/paths_tree.rs [hunk 8]
```diff
@@ -46,6 +48,89 @@ impl LookupResult {
...
+        .replace_bvars_with_fresh_fvars(|sort| rcx.define_var(sort))
+        .replace_generic_types(substs);
+
+    let constr = enum_constraint(&variant_def.ret, exprs);
+    rcx.assume_pred(constr);
+
+    variant_def.fields.to_vec()
+}
+
+fn downcast_place(
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-typeck/src/type_env/paths_tree.rs:113]: done!

File: flux-middle/src/global_env.rs [hunk 9]
```diff
@@ -80,13 +80,22 @@ impl<'genv, 'tcx> GlobalEnv<'genv, 'tcx> {
     }
 
     pub fn register_struct_def(&mut self, def_id: DefId, struct_def: core::StructDef) {
-        let variant = ty::conv::ConvCtxt::conv_struct_def(self, &struct_def);
+        let (adt_def, variant) = ty::conv::ConvCtxt::conv_struct_def(self, &struct_def);
         let variants = variant.map(|variant_def| vec![variant_def]);
         self.adt_variants.get_mut().insert(def_id, variants);
+        self.adt_defs.get_mut().insert(def_id, adt_def);
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-middle/src/global_env.rs:86]: ok removing that line (tests still pass!)

File: flux-middle/src/global_env.rs [hunk 10]
```diff
@@ -80,13 +80,22 @@ impl<'genv, 'tcx> GlobalEnv<'genv, 'tcx> {
...
+    pub fn register_enum_def(&mut self, def_id: DefId, enum_def: core::EnumDef) {
+        if let Some(variants) = ty::conv::ConvCtxt::conv_enum_def(self, enum_def) {
+            // println!("TRACE: register_enum_def {def_id:?} {variants:?}");
+            self.adt_variants.get_mut().insert(def_id, Some(variants));
+        }
+
+        let sorts = self.sorts_of(def_id);
+        self.adt_defs
+            .get_mut()
+            .insert(def_id, ty::AdtDef::new(self.tcx.adt_def(def_id), sorts));
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-middle/src/global_env.rs:98]: nuked

File: flux-typeck/src/checker.rs [hunk 11]
```diff
@@ -553,17 +554,21 @@ impl<'a, 'tcx, P: Phase> Checker<'a, 'tcx, P> {
...
+                    // like `Option` have no explicit downcast in MIR.
                     // Since places in mir have explicit downcast projecion we don't use this
-                    // extra control information and just assume places downcast to the correct variant
-                    // (guaranteed by rust type system). In the future we may explicitly track
-                    // the "active" variant.
+                    let tag = Tag::Goto(Some(src_info.span), target);
+                    let gen = &mut self.phase.constr_gen(self.genv, &rcx, tag);
+                    let mut down_place = place.clone();
+                    down_place.projection.push(PlaceElem::Downcast(variant_idx));
+                    env.lookup_place(&mut rcx, gen, &down_place);
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-typeck/src/checker.rs:571]: good idea, done

File: flux-middle/src/ty/conv.rs [hunk 12]
```diff
@@ -96,16 +97,45 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
...
+        enum_def: core::EnumDef,
+    ) -> Option<Vec<PolyVariant>> {
+        let mut cx = ConvCtxt::new(genv);
+        let variants: Vec<PolyVariant> = enum_def
+            .variants
+            .into_iter()
+            .map(|variant| cx.conv_variant(variant))
+            .collect();
+        if variants.is_empty() {
+            None
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-middle/src/ty/conv.rs:111]: TBH I've forgotten why I had this check - your point is good though, if you have a single `refined_by` you _have to_ provide the refined variants; have removed the `if variants.empty` ...

[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-middle/src/ty/conv.rs:111]: I take that back -- it _does_ break some of the tests where there are _no_ refined variants (but there are enums) e.g. 

```
 [ui] pos/structs/enum-join.rs
 [ui] pos/structs/enum-match-00.rs
 [ui] pos/structs/enum-match-03.rs
 [ui] pos/structs/simple-enum-constructor.rs
```

I've added a comment explaining the test.

**Comment by ranjitjhala** [comment]
@nilehmann -- I think I fixed the things you spotted earlier and added a couple of neg-tests too. LMK if you see anything else!

File: flux-middle/src/ty/conv.rs [hunk 13]
```diff
@@ -96,16 +97,49 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
...
+        let ret = self.conv_ty(&variant.ret, 1);
+        let variant = ty::VariantDef::new(fields, ret);
+        Binders::new(variant, sorts)
+    }
+
     pub(crate) fn conv_struct_def(
         genv: &GlobalEnv,
         struct_def: &core::StructDef,
-    ) -> Option<ty::VariantDef> {
+    ) -> (ty::AdtDef, Option<ty::PolyVariant>) {
```
Comment (nilehmann) [inline flux-middle/src/ty/conv.rs:136]: Since we removed the insertion of AdtDef in register_struct we don't have to return the AdtDef here.

**Review by nilehmann — APPROVED** [review]
@ranjitjhala lgtm! just a small detail.

**Pull request merged**